libfoundation/src/foundation-foreign.cpp

```diff
@@ -25,83 +25,176 @@
+MC_DLLEXPORT_DEF MCTypeInfoRef MCForeignCUIntTypeInfo() { return kMCCUIntTypeInfo; }
+MC_DLLEXPORT_DEF MCTypeInfoRef MCForeignCSIntTypeInfo() { return kMCCSIntTypeInfo; }
 MC_DLLEXPORT_DEF MCTypeInfoRef MCForeignCULongTypeInfo() { return kMCCULongTypeInfo; }
-MC_DLLEXPORT_DEF MCTypeInfoRef MCForeignCLongTypeInfo() { return kMCCLongTypeInfo; }
+MC_DLLEXPORT_DEF MCTypeInfoRef MCForeignCSLongTypeInfo() { return kMCCSLongTypeInfo; }
+MC_DLLEXPORT_DEF MCTypeInfoRef MCForeignCULongLongTypeInfo() { return kMCCULongTypeInfo; }
+MC_DLLEXPORT_DEF MCTypeInfoRef MCForeignCSLongLongTypeInfo() { return kMCCSLongTypeInfo; }
+
+/**/
 
 MC_DLLEXPORT_DEF MCTypeInfoRef MCForeignUIntTypeInfo() { return kMCUIntTypeInfo; }
-MC_DLLEXPORT_DEF MCTypeInfoRef MCForeignIntTypeInfo() { return kMCIntTypeInfo; }
+MC_DLLEXPORT_DEF MCTypeInfoRef MCForeignSIntTypeInfo() { return kMCSIntTypeInfo; }
 
 ////////////////////////////////////////////////////////////////////////////////
 
-struct bool_type_desc_t {
-    using c_type = bool;
-    static constexpr auto primitive_type = kMCForeignPrimitiveTypeBool;
-    static constexpr auto& base_type_info = kMCNullTypeInfo;
-    static constexpr auto& type_info = kMCBoolTypeInfo;
-    static constexpr auto is_optional = false;
-    static constexpr auto is_bridgable = true;
-    using bridge_type = MCBooleanRef;
-    static constexpr auto& bridge_type_info = kMCBooleanTypeInfo;
-    static constexpr auto& hash_func = MCHashBool;
+template<typename CType>
+constexpr MCForeignPrimitiveType __compute_signed_primitive_type()
+{
+    static_assert(sizeof(CType) != 1 || sizeof(CType) != 2 ||
+                  sizeof(CType) != 4 || sizeof(CType) != 8,
+                  "Unsupported signed integer size");
+    return sizeof(CType) == 1 ? kMCForeignPrimitiveTypeSInt8 :
+            sizeof(CType) == 2 ? kMCForeignPrimitiveTypeSInt16 :
+            sizeof(CType) == 4 ? kMCForeignPrimitiveTypeSInt32 :
+            sizeof(CType) == 8 ? kMCForeignPrimitiveTypeSInt64 :
+            kMCForeignPrimitiveTypeVoid;
+};
+
+template<typename CType>
+constexpr MCForeignPrimitiveType __compute_unsigned_primitive_type()
+{
+    static_assert(sizeof(CType) != 1 || sizeof(CType) != 2 ||
+                  sizeof(CType) != 4 || sizeof(CType) != 8,
+                  "Unsupported unsigned integer size");
+    return sizeof(CType) == 1 ? kMCForeignPrimitiveTypeUInt8 :
+            sizeof(CType) == 2 ? kMCForeignPrimitiveTypeUInt16 :
+            sizeof(CType) == 4 ? kMCForeignPrimitiveTypeUInt32 :
+            sizeof(CType) == 8 ? kMCForeignPrimitiveTypeUInt64 :
+            kMCForeignPrimitiveTypeVoid;
 };
 
+template<typename CType> constexpr MCForeignPrimitiveType
+__compute_primitive_type(void)
+{
+    static_assert(sizeof(CType) == 0, "No mapping from CType to primitive type");
+    return kMCForeignPrimitiveTypeVoid;
+}
+template<> constexpr MCForeignPrimitiveType
+__compute_primitive_type<char>(void) { return __compute_signed_primitive_type<char>(); }
```

Comment:
Unfortunately you are assuming that `char` is signed.

> The signedness of char depends on the compiler and the target platform: the defaults for ARM and PowerPC are typically unsigned, the defaults for x86 and x64 are typically signed.

I think it might be a good idea to explore the possibility of having two templates for `compute_primitive_type<T>()`, which use `enable_if<is_integral<T>::value && is_unsigned<T>::value>::type` to enable the correct one.  If that was possible, it would solve the "is `char` signed?" problem, and also might allow many of the explicit `compute_primitive_type<T>()` specialisations to be deleted.